File: array/132_pattern/solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
  bool find132pattern(vector<int>& nums) {
    int currK = INT_MIN;
    vector<int> s;

    for (int i = nums.size() - 1; i >= 0; i--) {
      if (nums[i] < currK) {
        return true;
      }

      while (s.size() > 0 && s.back() < nums[i]) {
        currK = max(currK, s.back());
        s.pop_back();
      }

      s.push_back(nums[i]);
    }
    return false;
  }
};

File: array/132_pattern/solution.rs
fn main() {
}

struct Solution {}

impl Solution {
  pub fn find132pattern(nums: Vec<i32>) -> bool {
    let mut curr_k = i32::MIN;
    let mut s: Vec<i32> = vec![];

    for i in (0..nums.len()).rev() {
      if nums[i] < curr_k {
        return true;
      }

      while s.len() > 0 && s.last().unwrap() < &nums[i] {
        curr_k = curr_k.max(nums[i]);
        s.pop();
      }

      s.push(nums[i]);
    }

    false
  }
}